man/infGather.1
INFGATHER(1)                User Manuals             INFGATHER(1)


NAME
    infGather - Script to gather information for a given IP or
		the given octect triple (i.e. 10.11.1). Or the
		user selected IP.

SYNOPSIS
    infGather [options...]

DESCRIPTION
     infGather gathers information about the given IP or octect
     triple. It uses nmap, dirb, nikto, enum4linux, etc. to 
     collect and capture useful information. The information is
     stored in the /$USER/Gather/program/IP/ directory, where 
     the program could be nmap/dirb/etc. By default, there are 
     configuration options that are set. Configuration options 
     like the word list to use, max time to sleep, etc. To 
     display more details use -h or --help. 

OPTIONS

     -f, --file		
          Read a file of IPs instead of scanning for alive boxes

     -h, --help		
          Help information

     -i, --ip		
          The IP triple to scan or the full IP to scan 
          (i.e. 10.11.1 OR 10.11.1.1)

     -m, --max		
          Maximum parallel processes to run 
          (Default is set to 10)

     -o, --only
          Runs only nmap scans. Quicker, since dirb can take
          a long time with larger word list files. Use this
          when you just want info about the target.

     -p, --perm		
          Install infGather so that it can be run from anywhere.
          Install man page and bash completion

     -q, --quick		
          Skip the heavier nmap scans

     -s, --single		
          Scan a single IP address

     -w, --wlist		
          Wordlist to use 
          (Default will be infGatherWordList.txt)

     -z, --ztime		
          Maximum sleep time between checking if PID completed
          (Default is set to 5s)

FILES
     /$USER/Gather/program/IP/
          The location of the collection of output from the given
          program. The program could be nmap, nikto, etc. For 
	  example, if it was nmap the directory location would 
          be /$USER/Gather/nmap/IP/. 

EXAMPLES 

     infGather -s 10.11.1.24
          Single scan of the IP 10.11.1.24, using the default
          word list.

     infGather -i 10.11.1 -w /usr/share/wordlists/dirb/common.txt
          Ping sweep to determine what IPs are "alive", scan the 
          "alive" IPs, using the word list /usr/share/wordlists/
          dirb/common.txt

     infGather -f ips.txt -m 20 -z 3
          Scan the IPs within the ips.txt file, setting max 
          parallel processes to 20, sleep time to 3s, using 
          the default word list.

REPORTING BUGS
     https://github.com/donutsThatsHowWeGetAnts/InfGather/bugs

AUTHOR
     donutsThatsHowWeGetAnts

Linux                Last change: MAY 2018                      1


bash/infGather.sh
#!/bin/bash
# --------------------------------------------------------
# Author: glender
# --------------------------------------------------------

OPTION=0
IO_FILE="$PWD/ips.txt"
declare -a pids
MAX_PARALLEL=12
MAX_SLEEP=5.0
WORDLIST=""

check_outfile()
{
	if [[ -f $IO_FILE && -s $IO_FILE ]]; then
		# clear the file
		> "${IO_FILE}"
	else
		touch "${IO_FILE}"
	fi
}

check_installs()
{
	if [ ! type nmap &> /dev/null ]; then
		echo "Please install nmap before you can proceed."
		exit 0
	fi

	if [ ! type enum4linux &> /dev/null ]; then
		echo "Please install enum4linux before you can proceed."
		exit 0
	fi
	
	if [ ! type nikto &> /dev/null ]; then
		echo "Please install nikto before you can proceed."
		exit 0
	fi

	if [ ! type dirb &> /dev/null ]; then
		echo "Please install dirb before you can proceed."
		exit 0
	fi
}

display_banner()
{
	echo ""
	echo ""
 	echo "###                #####                                    "
 	echo " #  #    # ###### #     #   ##   ##### #    # ###### #####  "
 	echo " #  ##   # #      #        #  #    #   #    # #      #    # "
 	echo " #  # #  # #####  #  #### #    #   #   ###### #####  #    # "
 	echo " #  #  # # #      #     # ######   #   #    # #      #####  "
 	echo " #  #   ## #      #     # #    #   #   #    # #      #   #  "
 	echo "### #    # #       #####  #    #   #   #    # ###### #    # "
 	echo ""
 	echo ""
	echo "Author: glender"
	echo "https://github.com/glender/InfGather"
	echo ""
	echo ""
}

display_usage()
{
	echo "Usage: ${0} [options...]"
	echo ""
        echo "		-f, --file		Read a file of IPs instead of scanning for alive boxes"
        echo "		-h, --help		Help information"
	echo "	[**]	-i, --ip		The IP triple to scan or the full IP to scan (i.e. 10.11.1 OR 10.11.1.1)"
	echo "		-m, --max		Maximum parallel processes to run (Default is set to $MAX_PARALLEL)"
	echo "		-s, --single		Scan a single IP address"
	echo "		-w, --wlist		Wordlist to use (Default will be infGatherWordList.txt)"
	echo "		-z, --ztime		Maximum sleep time between checking if PID completed (Default is set to $MAX_SLEEP s)"
	echo ""
	echo "	[**] is required to run the script"
	echo ""
}

check_ip()
{
	local ip=$1
	local split=(${ip//./ })

	# is the passed in arg of the format ###.###.###?
	if [[ $ip =~ ^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$ ]]; then
		# is each number less than or equal to 255?
		if [[ ${split[0]} -le 255 && ${split[1]} -le 255 && ${split[2]} -le 255 ]]; then
			echo -e "You provided a correct IP... \nMoving on... \n\n"
		else 
			echo -e "Incorrect IP! \nYour IP was out of range! \n\n"
			exit 0
		fi
	fi
}

check_ip_full()
{
	local ip=$1
	local split=(${ip//./ })

	# is the passed in arg of the format ###.###.###?
	if [[ $ip =~ ^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$ ]]; then
		# is each number less than or equal to 255?
		if [[ ${split[0]} -le 255 && ${split[1]} -le 255 && ${split[2]} -le 255 && ${split[3]} -le 255 ]]; then
			echo -e "You provided a correct IP... \nMoving on... \n\n"
			return 0
		else 
			echo -e "Incorrect IP! \nYour IP was out of range! \n\n"
			return 1
		fi
	fi
}

ping_sweep()
{
	check_outfile
	# iterate from 1..255
	for i in `seq 1 255`; do
		# ping to see if it is alive, save into temporary file
		ping -c 1 $1.$i | grep "bytes from" | cut -d " " -f 4 | cut -d ":" -f 1 >> $IO_FILE &
	done
}

check_running()
{
	while [ ${#pids[@]} -ge $MAX_PARALLEL ]; do
		for PID in "${pids[@]}"; do
			if [ -e /proc/${PID} ]; then
				# PID is still running
				continue
			else
				# the PID is no longer running, remove it
				delete=( $PID )
    				if [[ ${pids[PID]} = "${delete[0]}" ]]; then
      					unset 'pids[PID]'
    				fi

			fi
		done
	done
}

wait_for_completion()
{
	echo "Waiting for completion..."

	totalRunning=$(ps -C nmap -C nikto -C dirb -C enum4linux --no-headers | wc -l)

	sp="/-\|"
	i=1
	echo -n ' '
	until [ $totalRunning -eq 0 ]; do
		printf "\b${sp:i++%${#sp}:1}"
		sleep $MAX_SLEEP
		totalRunning=$(ps -C nmap -C nikto -C dirb --no-headers | wc -l)
	done

	reset

	# notify the user that we have completed 
	if [ ! type zenity &> /dev/null ]; then
		if [ type xmessage &> /dev/null ]; then
			xmessage -center 'message:$0 has completed. Please check your output.' 
		fi
	else
		echo 'message:$0 has completed. Please check your output.' | zenity --notification --listen
	fi
}

scans()
{
	location="/$USER/Gather/nmap/$1"
	mkdir -p $location

	today=`date '+%Y_%m_%d__%H_%M_%S'`
	nmap -n -sn -T4 $1 -o "$location/$1.$today.nmapHostDiscovery" > /dev/null &
	pids[$!]=$!
	check_running 

	nmap -sU -p 53 -sV -P0 --script dns-recursion $1 -o "$location/$1.$today.nmapDNSRecursion" > /dev/null &
	pids[$!]=$!
	check_running 

	nmap -Pn -p- -A $1 -r -n --open -o "$location/$1.$today.nmapAllOpenPorts" > /dev/null &
	pids[$!]=$!
	check_running 

	nmap -Pn -sL $1 -o "$location/$1.$today.nmapListScan" > /dev/null &
	pids[$!]=$!
	check_running

	nmap -sV -T4 --script vuln $1 -o "$location/$1.$today.nmapVulnScan" > /dev/null &
	pids[$!]=$!
	check_running

	nmap -Pn -sV -T4 -vv $1 -o "$location/$1.$today.nmapServiceDiscovery" > /dev/null &
	pids[$!]=$!
	check_running

	nikto_scan $! $location $1 &
	pids[$!]=$!
	check_running

	nmap -Pn -T4 -sU -vv $1 -o "$location/$1.$today.nmapUDPScan" > /dev/null &
	pids[$!]=$!
	check_running

	nmap -Pn -T4 -sS -p139,445 --script=smb-enum-users $1 -o "$location/$1.$today.nmapSmbEnumUsers" > /dev/null &
	pids[$!]=$!
	check_running
	
	nmap -Pn -T4 -sS -p139,445 --script=smb-enum-groups $1 -o "$location/$1.$today.nmapSmbEnumGroups" > /dev/null &
	pids[$!]=$!
	check_running
	
	nmap -Pn -T4 -sS -p139,445 --script=smb-enum-shares $1 -o "$location/$1.$today.nmapSmbEnumShares" > /dev/null & 
	pids[$!]=$!
	check_running

	dirb_scan $1 80 "0" &
	pids[$!]=$!
	check_running

	dirb_scan $1 443 "1" &
	pids[$!]=$!
	check_running

	enumlocation="/$USER/Gather/enum4linux/$1"
	mkdir -p $enumlocation
	enum4linux -U -o $1 > $enumlocation/$1.enum4linux 2>&1 &
	pids[$!]=$!
	check_running

}

nikto_scan()
{
	location="/$USER/Gather/nikto/$3"
	mkdir -p $location

	# wait for our main nmap scan to complete, our service detection
	while [ -e /proc/$1 ]; do
		sleep $MAX_SLEEP
	done

	# get ports other than the normal HTTP (80) and HTTPS (443) 
	httpPorts=$(grep -i http $2/*nmap* | grep open | grep -v "80/" | grep -v "443/" | awk -F: '{ print $2 }' | awk -F/ '{ print $1 }')
	
	# get the HTTPS ports
	httpsPorts=$(grep -i http $2/*nmap* | grep open | grep ssl | grep -v "80/" | grep -v "443/" | awk -F: '{ print $2 }' | awk -F/ '{ print $1 }')
	httpsPorts+=$(grep -i https $2/*nmap* | grep open | grep -v "80/" | grep -v "443/" | awk -F: '{ print $2 }' | awk -F/ '{ print $1 }')

	# ensure that they are unique entries
	httpPorts=($(echo "${httpPorts[@]}" | tr ' ' '\n' | sort -u | tr '\n' ' '))
	httpsPorts=($(echo "${httpsPorts[@]}" | tr ' ' '\n' | sort -u | tr '\n' ' '))

	# check the normal HTTPS port for our host
	nmap -p443 $3 -oG - | nikto -h - -output $location/nikto_https_$3.html > /dev/null &
	pids[$!]=$!
	check_running
	
	# check the normal HTTP port for our host
	nmap -p80 $3 -oG - | nikto -h - -output $location/nikto_http_$3.html > /dev/null &
	pids[$!]=$!
	check_running

	# remove the https ports from our http ports
	for target in "${httpsPorts[@]}"; do
  		for i in "${!httpPorts[@]}"; do
    			if [[ ${httpPorts[i]} = "${httpsPorts[0]}" ]]; then
      				unset 'httpPorts[i]'
    			fi
  		done
	done

	# process the https ports that we have found
	for port in "${httpsPorts[@]}"; do
		nikto -host $3 -port $port -ssl -output $location/nikto_https_$3_$port.html > /dev/null &
		pids[$!]=$!
		check_running
		dirb_scan $3 $port "0" &
		pids[$!]=$!
		check_running
	done

	# process the http ports that we have found
	for port in "${httpPorts[@]}"; do
		nikto -host $3 -port $port -output $location/nikto_http_$3_$port.html > /dev/null &
		pids[$!]=$!
		check_running
		dirb_scan $3 $port "1" &
		pids[$!]=$!
		check_running
	done

}

dirb_scan()
{
	location="/$USER/Gather/dirb/$1"
	mkdir -p $location

	# did the user supply a wordlist?
	if [ "$WORDLIST" == "" ]; then
		WORDLIST=$(find /$USER/ -name infGatherWordList.txt)
		# InfGather wordlist installed?
		if [ "$WORDLIST" == "" ]; then
			echo "Please install the InfGather wordlist, $USER."
			exit 0
		fi

	else
		# does the wordlist exist?
		if [ ! -f $WORDLIST ]; then
			echo "You supplied a wordlist that does not exist...shame on you!"
			exit 0
		fi
	fi

	# get the file size so that we can determine if we need to split into more than one file 
	filesize=$(stat -c%s $WORDLIST)
	base=$(dirname $WORDLIST) 
	fname=$(basename $WORDLIST)
	# half megabyte
	maxSize=500000 
	# newly created filename prefix
	prefix="qwaszx"
	created=0

	if (( filesize > maxSize )); then
		# split the file into parts
		pushd . > /dev/null
		cd $base
		split -l 5000 $fname $prefix
		popd > /dev/null
	fi

	files=$(shopt -s nullglob dotglob; echo $base/*)


	if [[ ${files[@]} =~ $prefix ]]; then
		files=( "${files[@]/$fname}" )
		files=( "${files[@]/$base}" )
		created=1
	fi

	if [ $created -eq 1 ]; then

		# iterate through our smaller files
		for f in $files; do

			if [[ $f =~ $prefix ]]; then
				# HTTP (0) or HTTPS (1)
				if [ "$3" -eq "0" ]; then
					dirb http://$1:$2/ $f -o $location/dirb_http_$2.txt -w > $location/dirb_http_$1_$2.txt &
					pids[$!]=$!
					check_running
				else
					dirb https://$1:$2/ $f -o $location/dirb_https_$2.txt -w > $location/dirb_httpd_$1_$2.txt &
					pids[$!]=$!
					check_running
				fi

				# clean up our created files
				rm -f $f
			fi
		done

	else

		# HTTP (0) or HTTPS (1)
		if [ "$3" -eq "0" ]; then
			dirb http://$1:$2/ $WORDLIST -o $location/dirb_http_$2.txt -w > $location/dirb_http_$1_$2.txt &
			pids[$!]=$!
			check_running
		else
			dirb https://$1:$2/ $WORDLIST -o $location/dirb_https_$2.txt -w > $location/dirb_httpd_$1_$2.txt &
			pids[$!]=$!
			check_running
		fi
	fi

}

read_file_and_scan()
{
	cat $IO_FILE | while read ip; do
		# is the IP address proper?
		check_ip_full $IP
		if [ $? -eq 0 ]; then
			scans $ip
		fi
	done
}

# display banner
display_banner

# check to make sure the user has the proper programs installed 
check_installs

# Get the command line options
for (( i=1; i<=$#; i++)); do
	case "${!i}" in
		# file to read IPs from
		-f|--file)
			arg=$((i+1))
			IO_FILE="${!arg}"
			OPTION=1
			shift # past - option
			;;
		# IP address to sweep
		-i|--ip)
			arg=$((i+1))
			IP="${!arg}"
			OPTION=2
			shift # past - option
			;;
		# single IP address scan
		-s|--single)
			arg=$((i+1))
			IP="${!arg}"
			OPTION=3
			shift # past - option
			;;
		# set maximum parallel processes
		-m|--max)
			arg=$((i+1))
			MAX_PARALLEL="${!arg}"
			shift # past - option
			;;
		# set the maximum sleep time between checking PID completion
		-z|--ztime)
			arg=$((i+1))
			MAX_SLEEP="${!arg}"
			shift # past - option
			;;
		-w|--wlist)
			arg=$((i+1))
			WORDLIST="${!arg}"
			shift # past - option
			;;
		# did the user ask for help?
		-h|--help)
			display_usage
			exit 0
			;;
		*)
			display_usage
			exit 0
			;;
	esac
done

# did the user supply the IP or the file?
if [ "$OPTION" -eq "0"  ]; then
	display_usage
	exit 0
else
	case "$OPTION" in
		"1")
			read_file_and_scan
			wait_for_completion
			;;
		"2")
			# is the IP address proper?
			check_ip $IP
			# sweep the IP address given
			ping_sweep $IP
			# wait for all of the spawned processes to finish
			wait
			read_file_and_scan
			wait_for_completion
			;;
		"3")
			# single IP
			check_ip_full $IP
			if [ $? -eq 0 ]; then 
				scans $IP
				wait_for_completion
			else
				exit 0
			fi
			;;
	
	esac
	
fi
